specs/HubProtocol.md

```diff
@@ -0,0 +1,225 @@
+# SignalR Hub Protocol
+
+The SignalR Protocol is a protocol for two-way RPC over any Message-based transport. Either party in the connection may invoke procedures on the other party, and procedures can return zero or more results or an error.
+
+## Terms
+
+* Caller - The node that is issuing an `Invocation` message and receiving `Result` messages (a node can be both Caller and Callee for different invocations simultaneously)
+* Callee - The node that is receiving an `Invocation` message and issuing `Result` messages (a node can be both Callee and Caller for different invocations simultaneously)
+* Binder - The component on each node that handles mapping `Invocation` messages to method calls and return values to `Result` messages
+
+## Transport Requirements
+
+The SignalR Protocol requires the following attributes from the underlying transport. The protocol was primarily designed for use with WebSockets, though it is relatively straightforward to build an adaptor layer for a different transport.
+
+* Message-based (aka Datagram, as opposed to Streaming)
+* A distinction between Text and Binary frames
+* Reliable, in-order, delivery of messages - Specifically, the SignalR protocol provides no facility for retransmission or reordering of messages. If that is important to an application scenario, the application must either use a transport that guarantees it (i.e. TCP) or provide their own system for managing message order.
+
+## Overview
+
+There are two encodings of the SignalR protocol: [JSON](http://www.json.org/) and [Protocol Buffers](https://developers.google.com/protocol-buffers/). Only one format can be used for the duration of a connection, and the format must be negotiated in advance (i.e. using a QueryString value, Header, or other indicator). However, each format shares a similar overall structure.
```

Comment:
OK. We could keep it closed for now and open later if needed. I don't really expect a lot of people asking for it but we will already have 2 implementations so we will probably have a design where things won't be hardcoded and therefore it should be easy to make them customizable if needed.